File: src/main/java/br/com/blz/testjava/business/ProductBusiness.java
package br.com.blz.testjava.business;

import br.com.blz.testjava.domain.Product;
import br.com.blz.testjava.enums.ErrorEnum;
import br.com.blz.testjava.exception.SkuExistsException;
import br.com.blz.testjava.exception.SkuNotExistsException;
import br.com.blz.testjava.service.ProductService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Objects;
import java.util.Optional;

@Service
@Slf4j
public class ProductBusiness {

    @Autowired
    private ProductService productService;

    public Product save(Product product) {
        if (Objects.nonNull(this.productService.findBySku(product.getSku())))
            throw new SkuExistsException(ErrorEnum.SKU_EXISTS.getDescription());
        return this.productService.save(product);
    }

    public Product findBySku(Long sku) {
        return Optional.ofNullable(this.productService.findBySku(sku))
            .orElseThrow(() -> new SkuNotExistsException(ErrorEnum.SKU_NOT_EXISTS.getDescription()));
    }
}

File: src/main/java/br/com/blz/testjava/product/business/ProductBusiness.kt
package br.com.blz.testjava.product.business

import br.com.blz.testjava.product.business.objects.Product
import br.com.blz.testjava.product.exception.ProductNotFoundException
import br.com.blz.testjava.product.exception.ProductSkuDuplicatedException
import br.com.blz.testjava.product.repository.ProductRepository
import org.springframework.stereotype.Service

@Service
class ProductBusiness() {
  fun create(product: Product) : Product = checkDuplicated(product)
    .also { ProductRepository.save(it) }

  private fun checkDuplicated(product: Product) : Product = ProductRepository.get(product.sku)
    ?.let { throw ProductSkuDuplicatedException(product.sku) }
    ?: product

  fun update(product: Product): Product = ProductRepository.get(product.sku)
    ?.let { ProductRepository.save(product) }
    ?: throw ProductNotFoundException(product.sku)

  fun get(sku: Long): Product = ProductRepository.get(sku) ?: throw ProductNotFoundException(sku)
}
